fix(image): dataUri is not random

OmkarPednekar opened the pull request:
If color option is not provided the code will use a randomly generated color code as default color
the fix(module):faker.image.dataUri commit is the final one. where I have used faker.color.rgb in the previous commit I was using a for loop to generate a hex code. @ST-DDT suggested me to use the inbuilt functions so that we have reproducible values.

<!-- Please run `pnpm run preflight` before opening a Pull Request to ensure that your code fulfills the minimal requirements for our project. -->

<!-- Please first read https://github.com/faker-js/faker/blob/next/CONTRIBUTING.md -->

<!-- Help us by writing a correct PR title following this guide: https://github.com/faker-js/faker/blob/next/CONTRIBUTING.md#committing -->

fixes #2307

Commit:
random color code gen added for image.DataUri

Commit:
fix(module):faker.image.dataUri

ST-DDT commented on src/modules/image/index.ts:
```diff
@@ -377,12 +377,16 @@ export class ImageModule {
+    const {
+      width = 640,
+      height = 480,
+      color = this.faker.color.rgb({ prefix: '#' }),
```
```suggestion
      color = this.faker.color.rgb(),
```

matthewmayer commented on src/modules/image/index.ts:
```diff
@@ -377,12 +377,16 @@ export class ImageModule {
        * The color of the image.
        *
-       * @default 'grey'
+       * @default 'random color'
```
this makes it look as if the default value is the literal string `'random color'`. I'd either change it to 

`
@default a random color
`

or omit the default value completely, and mention in the description that the color will be random unless specified.

OmkarPednekar replied:
Ok, will do the changes

OmkarPednekar replied:
![image](https://github.com/faker-js/faker/assets/74041929/047015d7-6bbb-4bd4-a66a-e67d97858b08)
 is this okay?

ST-DDT replied:
Other methods use:

````ts
@default faker.color.rgb()
````

ST-DDT commented:
I just noticed, that `dataUri` doesn't have seeded tests, so I created a PR for that.

Blocked by:

- https://github.com/faker-js/faker/pull/2317

OmkarPednekar commented:
is this okay? 
![image](https://github.com/faker-js/faker/assets/74041929/0bc122bc-5cb8-4a71-b5da-3c99eae994dc)

ST-DDT commented:
> is this okay? ![image](https://user-images.githubusercontent.com/74041929/260433226-0bc122bc-5cb8-4a71-b5da-3c99eae994dc.png)

I would probably phrase it like this:

`The background color of the image. Must be a color supported by svg. Defaults to a random color.`

OmkarPednekar commented:
Okay

Commit:
Description content changed

OmkarPednekar commented:
Please Check

ST-DDT commented on src/modules/image/index.ts:
```diff
@@ -375,17 +375,16 @@ export class ImageModule {
       height?: number;
       /**
-       * The color of the image.
+       * The color of the image. Must be a color supported by svg. Defaults to a random color.
```
Don't forget to change the `@param` docs above.

OmkarPednekar replied:
Done

Commit:
changes made to @params according to changes

OmkarPednekar commented:
Made the changes in @param as well

OmkarPednekar committed:
docs updated

OmkarPednekar commented:
![image](https://github.com/faker-js/faker/assets/74041929/e11c9b72-7cc5-4c24-967c-9e0678b3888f)
Updated the Docs

ST-DDT commented on src/modules/image/index.ts:
```diff
@@ -353,7 +353,7 @@ export class ImageModule {
    * @param options.width The width of the image. Defaults to `640`.
    * @param options.height The height of the image. Defaults to `480`.
-   * @param options.color The color of the image. Defaults to `grey`.
+   * @param options.color The color of the image. Defaults to a random color.
```
Both the `@param` description and the field description should be the same/similar.

```suggestion
   * @param options.color The color of the image. Must be a color supported by svg. Defaults to a random color.
```

ST-DDT replied:
Looks like you forgot this one. Or did you mark this as resolved intentionally?

OmkarPednekar replied:
Oh! So sorry I totally missed this one

OmkarPednekar replied:
The color of the image. Must be a color supported by svg. it should be this right? same as description

ST-DDT replied:
Plus the `Defaults to ...` part because that's not present in the `@param` block.

ST-DDT commented on src/modules/image/index.ts:
```diff
@@ -375,14 +375,18 @@ export class ImageModule {
       height?: number;
       /**
-       * The color of the image.
+       * The color of the image. Must be a color supported by svg. Defaults to a random color.
```
```suggestion
       * The color of the image. Must be a color supported by svg.
```

The `@default` already covers the `Defaults to` part.

OmkarPednekar replied:
okay

ST-DDT commented on test/modules/image.spec.ts:
```diff
@@ -471,8 +471,9 @@ describe('image', () => {
+      const expectedPattern = /^data:image\/svg\+xml;charset=UTF-8,/;
       const dataUri = faker.image.dataUri({ width: 200, height: 300 });
-      expect(dataUri).toMatchSnapshot();
+      expect(dataUri).toMatch(expectedPattern);
```
Please inline the expected pattern.

ST-DDT replied:
Maybe also check that the string contains the given width and height.

ST-DDT commented on test/modules/image.spec.ts:
```diff
@@ -471,8 +471,9 @@ describe('image', () => {
+      expect(dataUri).toMatch(expectedPattern);
     });
 
     it('should return a background color data URI', () => {
```
Maybe use the `expectedPattern` like check here as well (in addition to the snapshot)?

OmkarPednekar replied:
in this testcase 'should return a background color data URI' ?

OmkarPednekar replied:
![image](https://github.com/faker-js/faker/assets/74041929/7721162c-3cec-46fd-b2f6-0567ae06fa3d)
Made changes to the description and also added the testcase for checking width and height to both tests

OmkarPednekar committed:
Merge branch 'next' into next

OmkarPednekar committed:
Description Edited and Test for width and height +

OmkarPednekar committed:
Merge branch 'next' of https://github.com/OmkarPednekar/faker into next

ST-DDT commented on test/modules/image.spec.ts:
```diff
@@ -481,6 +482,8 @@ describe('image', () => {
         color: 'red',
       });
+      expect(dataUri).toMatch(/^data:image\/svg\+xml;charset=UTF-8,/);
+      expect(dataUri).toMatch(/width%3D%22200%22%20height%3D%22300/);
```
Don't forget to check for the color.

OmkarPednekar replied:
Added the check for color.

OmkarPednekar committed:
check added for color in dataUri String.

OmkarPednekar committed:
params updated

OmkarPednekar committed:
Merge branch 'next' into next

OmkarPednekar commented:
any updates on the review @matthewmayer

matthewmayer commented:
It's quite normal that PRs will take weeks or months before merging even if they are fine. Don't worry 😀

OmkarPednekar commented:
Okay Got It and also why is this one test failing?

ST-DDT commented:
> Okay Got It and also why is this one test failing?

Fixed

ST-DDT committed:
Merge branch 'next' into next

ST-DDT commented:
We added some (forgotten) seeded tests to ensure the values generated by this method are reproducible.
Please run `pnpm run preflight` to update the values.

OmkarPednekar committed:
Merge branch 'next' into next

OmkarPednekar commented:
> We added some (forgotten) seeded tests to ensure the values generated by this method are reproducible. Please run `pnpm run preflight` to update the values.

So i have to do this and then  commit again?

Commit:
"pnpm run preflight" executed

OmkarPednekar commented:
why are these checks failing?

ST-DDT commented:
> why are these checks failing?

The (new) seeded tests should highlight changed results caused by a refactoring/change.
This PR changes the behavior of the method (slightly) thus the  test snapshots needed to be updated.

ST-DDT committed:
Merge branch 'next' into next

OmkarPednekar committed:
Merge branch 'next' into next

ST-DDT committed:
Merge branch 'next' into next

OmkarPednekar committed:
Merge branch 'next' into next

OmkarPednekar committed:
Merge branch 'next' into next

xDivisionByZerox commented:
@OmkarPednekar Please resolve the merge conflicts if possible.

ST-DDT committed:
Merge branch 'next' into pr/OmkarPednekar/2316

ST-DDT committed:
Merge branch 'next' into next

ST-DDT commented:
I fixed the merge conflicts myself to speed up the merge process.

ST-DDT committed:
test: update snapshots

Shinigami92 commented:
👀

ST-DDT committed:
test: simplify random seeded tests

OmkarPednekar commented:
Why are these tests failing?

ST-DDT commented:
I fixed the random seeded tests using test snapshots.

ST-DDT commented:
> Why are these tests failing?

Because the randomized tests used snapshots which would work fine if the result is static, but this PR changes them to be random thus failing the tests. I fixed them by not using the static snapshots and instead check via validatorjs's isDataUri.

OmkarPednekar commented:
> > Why are these tests failing?
> 
> Because the randomized tests used snapshots which would work fine if the result is static, but this PR changes them to be random thus failing the tests. I fixed them by not using the static snapshots and instead check via validatorjs's isDataUri.

But I had removed those test cases and changed them to match the changes

ST-DDT commented:
> But I had removed those test cases and changed them to match the changes

https://github.com/faker-js/faker/pull/2273 added new ones.

- https://github.com/faker-js/faker/pull/2273

OmkarPednekar commented:
> > But I had removed those test cases and changed them to match the changes
> 
> #2273 added new ones.
> 
> * [feat(image): add image dataUri with base64 #2273](https://github.com/faker-js/faker/pull/2273)

Oh I seee so how can we fix this failing tests?

ST-DDT commented:
> Oh I seee so how can we fix this failing tests?

I already removed/fixed them.

OmkarPednekar commented:
> > Oh I seee so how can we fix this failing tests?
> 
> I already removed/fixed them.

No I meant the merge checks

ST-DDT commented:
> No I meant the merge checks

Not sure what you are referring to.

![grafik](https://github.com/faker-js/faker/assets/1579362/8188b54b-fe34-4d72-8e8c-d4da964b3f53)

OmkarPednekar commented:
> > No I meant the merge checks
> 
> Not sure what you are referring to.
> 
> ![grafik](https://user-images.githubusercontent.com/1579362/268393143-8188b54b-fe34-4d72-8e8c-d4da964b3f53.png)

Oh whatt just a moment ago they were appearing failed lol
